compiler/lcalc/optimizations.ml

```diff
@@ -26,20 +30,90 @@ let rec iota_expr (e : 'm expr) : 'm expr boxed =
   match Marked.unmark e with
   | EMatch { e = EInj { e = e'; cons; name = n' }, _; cases; name = n }
     when EnumName.equal n n' ->
-    let e1 = visitor_map iota_expr e' in
-    let case = visitor_map iota_expr (EnumConstructor.Map.find cons cases) in
-    Expr.eapp case [e1] m
+    (* match E x with | E y -> e1 = e1[y |-> x]*)
+    if true then
+      match Marked.unmark @@ EnumConstructor.Map.find cons cases with
+      (* holds because of invariant_match_inversion *)
+      | EAbs { binder; _ } -> visitor_map iota_expr (Expr.subst binder [e'])
+      | _ -> assert false
+    else
+      let e1 = visitor_map iota_expr e' in
+      let case = visitor_map iota_expr (EnumConstructor.Map.find cons cases) in
+      Expr.eapp case [e1] m
+      (* if the size of the expression is small enought, we just substitute. *)
   | EMatch { e = e'; cases; name = n }
-    when cases
-         |> EnumConstructor.Map.mapi (fun i case ->
-                match Marked.unmark case with
-                | EInj { cons = i'; name = n'; _ } ->
-                  EnumConstructor.equal i i' && EnumName.equal n n'
-                | _ -> false)
-         |> EnumConstructor.Map.for_all (fun _ b -> b) ->
+    when EnumConstructor.MapLabels.for_all cases ~f:(fun i case ->
+             match Marked.unmark case with
+             | EAbs { binder; _ } -> (
+               let var, body = Bindlib.unmbind binder in
+               (* because of invariant [invariant_match], the arity is always
+                  one. *)
+               let[@warning "-8"] [| var |] = var in
+
+               match Marked.unmark body with
+               | EInj { cons = i'; name = n'; e = EVar x, _ } ->
+                 EnumConstructor.equal i i'
+                 && EnumName.equal n n'
+                 && Bindlib.eq_vars x var
+               | EInj { cons = i'; name = n'; e = ELit LUnit, _ } ->
+                 (* since unit is the only value of type unit. We don't need to
+                    check the equality. *)
+                 EnumConstructor.equal i i' && EnumName.equal n n'
+               | _ -> false)
+             | _ ->
+               (* because of invariant [invariant_match], there is always some
+                  EAbs in each cases. *)
+               assert false) ->
     visitor_map iota_expr e'
   | _ -> visitor_map iota_expr e
 
+let rec iota2_expr (e : 'm expr) : 'm expr boxed =
```

Comment:
What is the difference between `iota_expr` and `iota2_expr`?

Reply:
I've added a comment about iota_expr in the mli file.